### ISSUE TITLE
[RFC] re-export the oci-spec from `libcontainer`

### ISSUE THREAD

**Issue by yihuaf** [containers/youki#2066]
As a user of `libcontainer`, having to match `libcontainer` version with the `oci-spec-rs` version is a pain. The `oci-spec-rs` can introduce breaking changes. For example, with the introduction of time namespace, the libcontainer is broken and require code change. In this case, the users need to spend efforts trying to understand what happens.

Making oci-spec-rs backward compatible does not make much sense. Therefore I propose we re-export the oci-spec-rs that `libcontainer` uses. In this way, to use `libcontainer`, users don't have to import `oci-spec-rs` into their own dependencies.

Ref: https://www.lurklurk.org/effective-rust/re-export.html

**Comment by YJDoc2** [comment]
Hey, even though the overall idea itself is good, it will mean we need to do a breaking release of libcontainer every time we update the oci spec, even if we changed only internal use and not external API. (That said semvar with such kind of changes does get subtle and complex, so maybe we should do breaking releases?)
Also,
- can you elaborate more on the breakage you mentioned? I'm not familiar with it :sweat_smile: 
- what would be effect of trying to make oci-spec backword compatible? It won't be simple, but if it can prevent breakages, it might be useful.

Overall, if there is no better option, I think we should re-export the oci-spec as you have mentioned.

**Comment by yihuaf** [comment]
> can you elaborate more on the breakage you mentioned? I'm not familiar with it 

oci-spec introduced the `time` namespace from 0.6.0 to 0.6.1. As a result, it broke libcontainer because we have to exhaustively matching all namespace type. For us, this is OK because we would just update both `youki` and `libcontainer` at the same time. However, for people who use only `libcontainer`, they will have to include `oci-spec` in their dependency to use `libcontainer` because `libcontainer` uses types from `oci-spec` in its API. One of the solution is to re-export the types used in the API as part of the crate.

> what would be effect of trying to make oci-spec backword compatible? It won't be simple, but if it can prevent breakages, it might be useful.

The main issue is that oci-spec is a spec. It has to add new things. For example, with the introduction of a new namespace, there is no easy way to make this backward compatible. The easiest and the most nature way is to add to the namespace type enum. There are ways to work around the issue, but this is just one example. It is far simpler to make the changes to the spec, bump the version, and let `libcontainer` and `youki` to release a new version implementing the new spec.

Plus, oci-spec follows the oci runtime spec. It can be a hard sell trying to coordinate runc, crun, and the oci runtime spec to all trying to respect backward compatibility.

> it will mean we need to do a breaking release of libcontainer every time we update the oci spec, even if we changed only internal use and not external API. 

Can you give an example of the breaking release? libcontainer version is pinned to a specific commit, which has a specific oci-spec version. When oci-spec version changes, we change in libcontainer. We don't have to release libcontainer because the old libcontainer version will still use the old oci-spec.

In fact, what you are describing is how it works today. I point to the example above. When oci-spec is bumped from 0.6.0 -> 0.6.1, the existing libcontainer breaks. We will have to release a new version of the libcontainer to fix the breakage. This creates a mis-match between the version of the oci-spec and the version of libcontainer. This RFC, using re-export, avoids exactly this issue. Now, the changes in oci-spec doesn't affect libcontainer at all, until we release a new version of libcontainer with the new oci-spec version. In the PR I linked, see how `youki` crate now has no dependency on the `oci-spec` crate anymore. The dependency is solely within the `libcontainer` crate.

**Comment by utam0k** [comment]
First of all, the largest user of oci-spec-rs is already not libcontianer. So this problem is not only ours.
It is also dangerous to change repositories. Some people may refer to it directly.

The problem would occur to those who use libcontainer and oci-spec-rs directly.
However, I think this is the responsibility of the user. Wouldn't it be better to specify a minimum version of ocio-spec-rs instead of a fixed version of ocio-spec-rs?
https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#comparison-requirements

The same should be done for libcontainer, which should state the minimum version to be supported instead of a fixed version.

**Comment by utam0k** [comment]
I didn't know there was such a thing as re-export. Dare I ask what the disadvantages of this are?

**Comment by yihuaf** [comment]
I believe this comment from the rust-lang discussion summarizes the issue much better than I did above: https://github.com/rust-lang/api-guidelines/discussions/176#discussioncomment-232636

I don't think re-export itself has a clear down side. This technique is the right choice when the library's public api depends on another 3rd party crate. In our case, that is libcontainer's api depends on oci-spec-rs. As a result, any breakage in oci-spec-rs will break libcontainer. I don't think we can avoid using oci-spec in the libcontainer's public api.

I came across a reddit thread with one con. If we strictly follows semvar, which I don't think we are at the moment, a major version bump would require us to bump the major version as well. Semvar can be used to indicate api breakage, but we are not using it. For example, 0.6.0 should really be 0.7.0 instead of 0.6.1. A patch version bump signals to users that there is typically no breakage. But again, not everyone follows this and many just use semvar as a numbering scheme. Ref: https://www.reddit.com/r/rust/comments/b7n5ue/is_it_ok_to_publicly_reexport_types_from_3rd/

In fact, this is a proposal to make re-export of dependencies in the public interface a lint rule: https://github.com/rust-lang/rust-clippy/issues/2848

In the end, I don't think this is something controversial. We choose to use oci-spec-rs as a separate crate in our api, so we already made our decisions.
